src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -356,27 +321,82 @@ private void handleSetAllowedSites(RoutingContext rc) {
+                if(newKeyset == null) return;
+                JsonObject jo = new JsonObject();
+                jo.put("allowed_sites", allowedSites);
+                jo.put("allowed_types", newKeyset.getAllowedTypes());
+                jo.put("hash", newKeyset.hashCode());
+
+                rc.response()
+                        .putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                        .end(jo.encode());
             } catch (Exception e) {
                 rc.fail(500, e);
-                return;
             }
+        }
+    }
 
-           JsonObject jo = new JsonObject();
-           jo.put("allowed_sites", allowedSites);
-           jo.put("hash", newKeyset.hashCode());
+    private AdminKeyset setAdminKeyset(RoutingContext rc, JsonArray allowedSites, JsonArray allowedTypes,
+                                       Integer siteId, Integer keysetId, String name)
+            throws Exception{
 
-            rc.response()
-                    .putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
-                    .end(jo.encode());
+        if (keysetId == null) {
+            keysetId = this.keysetManager.getNextKeysetId();
+            name = "";
+        }
+
+        final Set<Integer> newlist;
+
+        if (allowedSites != null){
+            OptionalInt firstInvalidSite = allowedSites.stream().mapToInt(s -> (Integer) s).filter(s -> siteProvider.getSite(s) == null).findFirst();
+            if (firstInvalidSite.isPresent()) {
+                ResponseUtil.error(rc, 400, "Site id " + firstInvalidSite.getAsInt() + " not valid");
+                return null;
+            }
+
+            boolean containsDuplicates = allowedSites.stream().distinct().count() < allowedSites.stream().count();
+            if (containsDuplicates) {
+                ResponseUtil.error(rc, 400, "Duplicate site_ids not permitted");
+                return null;
+            }
+
+            newlist = allowedSites.stream()
+                    .mapToInt(s -> (Integer) s)
+                    .filter(s -> !Objects.equals(s, siteId))
+                    .boxed()
+                    .collect(Collectors.toSet());
+        } else {
+            newlist = null;
         }
+
+        Set<ClientType> newAllowedTypes = null;
+        if(allowedTypes == null || allowedTypes.isEmpty()) {
+            newAllowedTypes = new HashSet<>();
+        } else {
+            newAllowedTypes = allowedTypes.stream()
+                    .map(s -> Enum.valueOf(ClientType.class, s.toString()))
+                    .collect(Collectors.toSet());
+        }
+
+        final AdminKeyset newKeyset = new AdminKeyset(keysetId, siteId, name,
+                newlist, Instant.now().getEpochSecond(), true, true, newAllowedTypes);
+
+        this.keysetManager.addOrReplaceKeyset(newKeyset);
+        return newKeyset;
+    }
+
+    private JsonObject jsonFullKeyset(AdminKeyset keyset) {
+        JsonObject jo = new JsonObject();
```

Comment:
NP: I know we already have a lot of similar stuff so will approve this MR, but for new stuff let's try to use "proper" serialization/deserialization with Jackson as described eg here https://www.baeldung.com/jackson-deserialization. 
Based on [this discussion ](https://thetradedesk.slack.com/archives/C050J27TL4B/p1692752400935869)in foundations (also including Jonathan and Matt) added Dealing with Json into team code conventions: https://atlassian.thetradedesk.com/confluence/display/UID2/UID2+Java+Coding+Guidelines+and+Design+Conventions

Reply:
https://atlassian.thetradedesk.com/jira/browse/UID2-1730